In [2]:
import nltk
import re

In [ ]:
"""" Basic Regular Expression Meta-Characters, Including Wildcards, Ranges and Closures
.	        Wildcard, matches any character
^abc	    Matches some pattern abc at the start of a string
abc$	    Matches some pattern abc at the end of a string
[abc]	    Matches one of a set of characters
[^abc]      Matches anything but a set of characters
[A-Z0-9]	Matches one of a range of characters
ed|ing|s	Matches one of the specified strings (disjunction)
*	        Zero or more of previous item, e.g. a*, [a-z]* (also known as Kleene Closure)
+	        One or more of previous item, e.g. a+, [a-z]+
?	        Zero or one of the previous item (i.e. optional), e.g. a?, [a-z]?
{n}	        Exactly n repeats where n is a non-negative integer
{n,}	    At least n repeats
{,n}	    No more than n repeats
{m,n}	    At least m and no more than n repeats
a(b|c)+	    Parentheses that indicate the scope of the operators


# regular expression 
* list가 아니라 string에만 적용된다.
* ( ) 안에 들어 있는 것이 입력


In [22]:
re.search('abc', 'abced')

<_sre.SRE_Match object; span=(0, 3), match='abc'>

* 입력이 2개인 함수이다.
* abcde라는 것에서 abc를 찾아라.

In [5]:
wordlist = [w for w in nltk.corpus.words.words('en') if w.islower()]
print(wordlist[:10])


['a', 'aa', 'aal', 'aalii', 'aam', 'aardvark', 'aardwolf', 'aba', 'abac', 'abaca']


* wordlist는 string들의 list
* re.search는 list가 아니라 string에만 적용되기 때문에 for루프를 돌려서 하나씩 넣어주어야 한다.

In [6]:
result = [w for w in wordlist if re.search('ed$', w)][:10]
print(result[:10])

['abaissed', 'abandoned', 'abased', 'abashed', 'abatised', 'abed', 'aborted', 'abridged', 'abscessed', 'absconded']


* '~$'은 ~으로 끝나는 것을 의미한다.
* 이 경우 ed로 끝나는 단어를 앞에서부터 10개 보여주라는 의미

In [ ]:
result = [w for w in wordlist if re.search('^..j..t..$', w)][:10]
print(result[:10]

* .에는 어떤 글자라도 들어갈 수 있다.
* 이 경우 3번째 글자가 j, 5번째 글자가 t인 8글자의 단어를 앞에서부터 10개 보여주라는 의미

In [7]:
result = [w for w in wordlist if re.search('^[ghi][mno][jlk][def]$', w)][:10]
print(result[:10])

['gold', 'golf', 'hold', 'hole']


* []은 안에 들어간 것 중에서 하나를 의미한다. []으로 묶으면 그 중 무엇이든 상관 없다는 뜻
* matches one of a set of characters. 
* 이 경우에는 g/h/i중에 하나, m/n/o중에 하나, j/l/k중에 하나, d/e/f중에 하나 인 4자리 글자를 보여주면 된다.
* 앞에서부터 10개 표시하라고 하였으나, 조건에 해당하는 글자가 4개뿐이라 4개만 나옴.

In [8]:
result = [w for w in wordlist if re.search('^[ah]+$', w)][:10]
print(result[:10])

['a', 'aa', 'ah', 'aha', 'h', 'ha', 'hah']


* +는 해당 글자가 한 번 이상 반복되는 것을 의미한다.
* 이 경우에는 a/h가 한 번 이상 반복되는 것을 보여주라는 의미.

In [9]:
wsj = sorted(set(nltk.corpus.treebank.words()))
wordlist = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
print(wordlist[:10])

['0.0085', '0.05', '0.1', '0.16', '0.2', '0.25', '0.28', '0.3', '0.4', '0.5']


* -는 사이에 있는 것 모두, 즉 0-9는 0,1,2,3,4,5,6,7,8,9를 말한다. 
* 이 경우 조건은 모든 소수를 의미할 것.

In [11]:
wordlist = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
print(wordlist[:10])

['C$', 'US$']


* $는 ~로 끝나는 것을 의미한다고 말했는데, 정말 글자 그대로의 $가 쓰고 싶다면 앞에\를 붙이면 된다.
* 이 경우에는 대문자 알파벳으로 시작하고, $로 끝나는 것을 보여주라는 의미 

In [13]:
wordlist = [w for w in wsj if re.search('^[0-9]{4}$', w)]
print(wordlist[:10])

['1614', '1637', '1787', '1901', '1903', '1917', '1925', '1929', '1933', '1934']


In [15]:
wordlist = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]
print(wordlist[:10])

['black-and-white', 'bread-and-butter', 'father-in-law', 'machine-gun-toting', 'savings-and-loan']


In [16]:
wordlist = [w for w in wsj if re.search('(ed|ing)$', w)]
print(wordlist[:10])

['62%-owned', 'Absorbed', 'According', 'Adopting', 'Advanced', 'Advancing', 'Alfred', 'Allied', 'Annualized', 'Anything']


In [17]:
word = 'supercalifragilisticexpialidocious'
result = re.findall(r'[aeiou]', word)
print(result)

['u', 'e', 'a', 'i', 'a', 'i', 'i', 'i', 'e', 'i', 'a', 'i', 'o', 'i', 'o', 'u']


In [18]:
result = re.findall('[aeiou](..)[aeiou]', word)
print(result)

['rc', 'fr', 'st', 'xp', 'ci']


* 주어진 word에 모음과 모음 사이에 자음 두 글자가 들어가 있다면, 이 자음 두 개의 조합을 모두 보여주라는 의미

In [19]:
result = re.findall('[^aeiou].+[^aeiou]', word)
print(result)

['supercalifragilisticexpialidocious']


In [20]:
result = re.findall('[^aeiou].+?[^aeiou]', word)
print(result)

['sup', 'rcal', 'frag', 'lis', 'tic', 'xpial', 'doc']


In [21]:
wsj = sorted(set(nltk.corpus.treebank.words()))
fdist = nltk.FreqDist(vs for word in wsj for vs in re.findall(r'[aeiou]{2,}', word))
fdist.most_common(10)

[('io', 549),
 ('ea', 476),
 ('ie', 331),
 ('ou', 329),
 ('ai', 261),
 ('ia', 253),
 ('ee', 217),
 ('oo', 174),
 ('ua', 109),
 ('au', 106)]

* findall은 뒤의 string에서 있는 것을 다 찾아라
* Freqdist의 variable type은 list. 
* [aeiou]{2,} = aeiou가 2개 이상 들어 있기만 하면 그것을 찾아와서 list로 만들어라. 
* 그 list에서 fdist를 받고 . most_common 하면, 상위 10개를 보여준다.
* 'io', 549 는 io가 549번 나와서 가장 자주 나왔다는 것.

In [3]:
regexp = r'^[AEIOUaeiou]+|[AEIOUaeiou]+$|[^AEIOUaeiou]'
def compress(word):
    pieces = re.findall(regexp, word)
    return ''.join(pieces)

english_udhr = nltk.corpus.udhr.words('English-Latin1')
print(nltk.tokenwrap(compress(w) for w in english_udhr[:75]))


    a   e   i   o   u 
k 418 148  94 420 173 
p  83  31 105  34  51 
r 187  63  84  89  79 
s   0   0 100   2   1 
t  47   8   0 148  37 
v  93  27 105  48  49 


In [27]:
rotokas_words = nltk.corpus.toolbox.words('rotokas.dic')
cvs = [cv for w in rotokas_words for cv in re.findall(r'[ptksvr][aeiou]', w)]
cfd = nltk.ConditionalFreqDist(cvs) # CFD works pairwise only
cfd.tabulate()

    a   e   i   o   u 
k 418 148  94 420 173 
p  83  31 105  34  51 
r 187  63  84  89  79 
s   0   0 100   2   1 
t  47   8   0 148  37 
v  93  27 105  48  49 


In [ ]:
cv_word_pairs = [(cv, w) for w in rotokas_words for cv in re.findall(r'[ptksvr][aeiou]', w)]
cv_index = nltk.Index(cv_word_pairs)
cv_index['po']